### ISSUE TITLE
C structures using typedef don't recieve proper documentation.

### ISSUE THREAD

**Issue by Livingwind** [jnikula/hawkmoth#11]
The autodoc feature doesn't properly represent structures of the form
`typedef struct {...} <symbol>;` which is a common way of defining structures in C code.

A temporary solution was to write `typedef struct <symbol> {...} <symbol>;` but this is redundant.

**Comment by jnikula** [comment]
Thanks for the report. I think @BrunoMSantos had code to address this kind of issues, but I'm afraid we've stalled on it a bit. It's definitely something we need to address.

**Comment by BrunoMSantos** [comment]
Indeed.

@Livingwind, @mv0, could I interest you in demoing the `anonym` branch from my [fork](https://gitlab.com/bms-contrib/hawkmoth/tree/anonym)? Basically anything involving anonymous structures, unions and enumerators should be scrutinized. Once we know we know what the correct behaviour is, it shouldn't be too complicated to implement it in a mergeable way.

There's also a long, but stale discussion about this in the mailing list. Starts out in https://www.freelists.org/post/hawkmoth/PATCH-v2-1215-hawkmoth-fix-documentation-of-anonymous-types
Both the email thread and the examples and tests in that branch should offer a glimpse of our reasoning so far.

**Comment by Livingwind** [comment]
I'm planning on approaching this more this weekend but I'd love to get the conversation moving again.

Reading through the mailing list, it sounds like @jnikula reached a point where there wasn't a good solution to named/typedef'ed structs. Specifically in reference to [this email](https://www.freelists.org/post/hawkmoth/PATCH-v2-1215-hawkmoth-fix-documentation-of-anonymous-types,4), would it be possible to add:
- case where struct is named but not typedef'ed: `/** document struct name */`
- case where struct is named w/ usable type: `/** document struct name :: types: sym1, sym2... */` or have this be a proper of the named structure rather than an alternative.
- case where struct is anon w/ usable type: `/** documents sym1, sym2 ... */`
It's important to note that there can be an arbitrary number of types defined, not just one like in the examples @jnikula provided in the mailing list.

While the behavior is fairly complicated I think it hits the 3 big ways people like to write structure definitions and matches expected behavior for each. Strictly speaking about the implementation, you can always look for declared symbols after a struct and use them in place of names when the name isn't defined.

Both of you also mentioned Doxygen compatibility when approaching this issue but it would be nice to implement a feature that isn't currently in Doxygen and treat that as a selling point for the library (and default back to Doxygen behavior if compatibility was a core feature).

I'm in the process of coercing people to provide better documentation so it's important that the tooling does't enforce a specific way to write code.

**Comment by BrunoMSantos** [comment]
> Reading through the mailing list, it sounds like @jnikula reached a point where there wasn't a good solution to named/typedef'ed structs.

If I remember correctly, we actually agreed there, but we need to cover all the cases carefully, not just from the perspective of what 'works', but even more important what 'currently works, but really shouldn't'.

Overall, we really must go for least surprise here. And in some cases, that means the user must name structures and separate the instances.

> While the behavior is fairly complicated I think it hits the 3 big ways people like to write structure definitions and matches expected behavior for each.

I have to admit that I didn't really understand your cases. Could you write them as short snippets?

> Strictly speaking about the implementation, you can always look for declared symbols after a struct and use them in place of names when the name isn't defined.

Ah, not as simple as it may sound :)

Clang's AST is (understandably) built for translation of source into machine code and all that it entails. Traversing it in some directions is easy (having a certain instance, find its type), but it's not so easy the other way around (having a type, find all the instances of that type). Let's face it, when would a compiler or similar tool ever need to do that? Sadly, we need it if we are to catch some edge cases here, both to issue documentation and (maybe more so) errors.

This means that for anything but the most trivial implementations (like my branch), we need to add our own metadata when parsing the code (what we refer to as a 2nd/extra pass in the emails), effectively building a 'custom AST' we can then use in the last `parse` step. Gut feeling is that this is no small task, but haven't spent much time on it.

> Both of you also mentioned Doxygen compatibility when approaching this issue but it would be nice to implement a feature that isn't currently in Doxygen and treat that as a selling point for the library (and default back to Doxygen behavior if compatibility was a core feature).

This relates to Doxygen's poor handling (in my view) of anonymous structures which I'm using as a frame of reference. That branch already does better than Doxygen as far as I'm concerned. But I think we all agree, absolutely no one is making a case for Doxygen compatibility ;)

**Comment by Livingwind** [comment]
An example of the 3 ways people write structs I can think of off the top of my head:
Case 1:
```
typedef struct Name {
  members...
};
```
which results in /** documents struct Name */
Case 2:
```
typedef struct Name {
  members...
} my_name, your_name;
```
which results in /** documents struct Name :: my_name, your_name */ (or some way of showing the typedef)
Case 3:
```
typedef struct {
  members...
} my_name, your_name;
```
which results in /** documents struct my_name, your_name */

I don't have your understanding of Clang AST so I can't exactly comment on the difficulty but I was trying to respond where in your email you said:
/** documents struct named, **not type_t** */
        typedef struct named { ... } type_t;
The name of the above struct is 'named' and that should be what is referenced in the documentation. I do however think 'type_t' should be referenced somewhere in the documentation for 'named'.

**Comment by BrunoMSantos** [comment]
Note that Sphinx directives work with a single symbol. So consider this example:

```c
/** Description of a *and* b. */
struct {
        /** Some member. */
        int member;
} a, b;
```

This is the 'least surprising' thing we could possibly generate:

```rst
.. c:var:: <anonymous> a, b
   Description of a *and* b.

   .. c:member:: member
      Some member.
```

or maybe for another `c:var::` implementation: 

```rst
.. c:var:: <anonymous> a, <anonymous> b
   Description of a *and* b.

   .. c:member:: member
      Some member.
```

Implementing this wouldn't be trivial on our side, but the end result looks pretty good in reST, problem is that this is not the `c:var::` syntax, which would interpret (e.g.) the 1st case as a single variable `b` of type `<anonymous> a,`. This breaks the formatting and reference tags in the compiled documentation. This leaves us with massive duplication as an alternative, which is horrible:

```rst
.. c:var:: <anonymous> a
   Description of a *and* b.

   .. c:member:: member
      Some member.

.. c:var:: <anonymous> b
   Description of a *and* b.

   .. c:member:: member
      Some member.
```

This is why we were aiming at something much simpler, supporting only cases where there's a single instance of an anonymous something.

----------------------------------

To complicate matters further though, I realized that there are scoping issues with `c:member::` which is only supposed to be placed inside of `c:type::`, not `c:var::`. The take away is that we need some features upstream:

1. Let `c:var::`, `c:type::` and `c:member::` take a comma separated list of symbols.

2. `c:member::` must issue as many back references to each member as there are variables.

3. Let `c:var::` behave like `c:type::` for scoping of `c:member::`.
There's a dirty workaround by the way, but I'd rather not go there as it relies on what is likely undefined behaviour or just a plain bug in Sphinx's C domain.
This is needed for even the simple cases supported in `anonym` branch.

So with even the `anonym` branch currently broken, it doesn't look like we can support any anonymous stuff for the time being :/

@jnikula, am I making any sense?

**Comment by jnikula** [comment]
I'll start off with listing cases that I think we should **not** put effort into, and which I think can be reasonably worked around by the user:

* Mixed type and instance declarations. For example, `struct typename { ... } varname;`.

* Multiple type definitions. For example, `typedef int typename1, typename2;`.

* Multiple variable declarations. For example, `int varname1, varname2;`.

* Documentation of the anonymous types in variable declarations. For example, `struct { ... } varname;` or `enum { ... } varname;`. These should turn into simple variable documentation with no regard to the type.

For all of these, my argument is the same. While they're all valid and commonly used constructs, I think if they deserve to be documented, they deserve to be declared separately with documentation comments of their own. We can add support for these later on if it turns out to be simple, but for now I wouldn't worry about them at all.

Typedefs of named and unnamed structs, unions and enums is another matter. For example, `typedef struct { ... } typename;` and `typedef enum name { ... } typename;` are something I think we need to focus on. And I think the typedef for the unnamed ones should turn into proper documentation for the struct, union or enum, with the given typename, and I think this should have priority over the named one. And this goes the full circle back to the original report here.

**Comment by BrunoMSantos** [comment]
Nice summary. Yeah, I'm also more inclined to do this sort of thing for its much greater simplicity and for not requiring features 1. and 2. from my list, which may not even be well received upstream to begin with.

> Documentation of the anonymous types in variable declarations. For example, struct { ... } varname; or enum { ... } varname;. These should turn into simple variable documentation with no regard to the type.

So this
```c
/** This is a variable. */
struct {
        /** Member. */
        int foo;
} bar;
```
should yield this?
```rst
.. c:var:: struct <anonymous> bar
   This is a variable.
```

Granted we'd not be affected by the missing 3rd feature either and it simplifies it even further. Of course some cases I had intended to support are no longer possible, like documenting the whole `struct foo`, including the nested `bar` without declaring the internal structure separately:

```c
/** This is a structure which has another structure inside. */
struct foo{
        /** Member structure. */
        struct {
                /** Nested something. */
                int bar;
        } baz;
};
```

Of course there are some merits to this approach... maybe we shouldn't care about 2nd+ level members and they should be named and documented separately if relevant. I can be convinced of doing it this way.

**Comment by jnikula** [comment]
@Livingwind What's your original expectation?

We have a test case which expects https://github.com/jnikula/hawkmoth/blob/master/test/typedef-struct.c to generate https://github.com/jnikula/hawkmoth/blob/master/test/typedef-struct.rst.

**Comment by Livingwind** [comment]
For those cases, that's what I was expecting. I'd just be nice to somewhere in the rst file reference the actual typedef'ed value (`typedef_struct_named` for `struct named` in the file you linked above).

I also don't see the unnamed typedef'ed struct working in my version of Hawkmoth (which I installed from PyPi). Do you plan on adding these changes to PyPi soon?

**Comment by jnikula** [comment]
> I also don't see the unnamed typedef'ed struct working in my version of Hawkmoth (which I installed from PyPi). Do you plan on adding these changes to PyPi soon?

That's odd. I don't think we've made changes to this part in a long time. Can you provide a source code example that fails?

**Comment by Livingwind** [comment]
Sure, I currently have Hawkmoth 0.3 installed using Sphinx 1.8
```
/** Description */
typedef struct {
   /** Input/Output Buffer */
   void           *mptr;
} msg_t;
```

**Comment by jnikula** [comment]
If I save that to `test.c` and run `python3 -m hawkmoth test.c` I get:

```
.. c:type:: msg_t

   Description


   .. c:member:: void * mptr

      Input/Output Buffer
```

Do you get something else, or is that not the output you expect?

**Comment by Livingwind** [comment]
Yeah I get the same results when I run the command you provided. I did however remove a lot of my code for privacy reasons. Here is a better example...

```
/** Description */
typedef struct {
    /** msg pointer */
    SomeStruct mptr;
} msg_wrapper;
```
Generates:
```
.. c:type:: struct (anonymous at main.c:3:9)

   Description


   .. c:member:: int mptr

      msg pointer
```
When the parser encounters an undefined member, the name of the struct becomes anonymous.

**Comment by jnikula** [comment]
Ahah! The compiler must have knowledge about the types it encounters, including struct members. We get incomplete information because the compilation fails.

We just gained the ability to log compiler diagnostics, so I did have the benefit of this output with current master: `ERROR: test.c:4: unknown type name 'SomeStruct'`

It is a shortcoming in our documentation that we don't emphasize you likely need to pass the `CFLAGS` to clang via either `cautodoc_clang` config option or `clang` option to the `c:autodoc` directive in order for the build to work.

**Comment by jnikula** [comment]
Oh, and we probably need to look at more practical ways to pass the `CFLAGS` via `sphinx-build` command line.

**Comment by Livingwind** [comment]
Is there no way to just print the symbol even though the compiler fails? It seems like a big pain to add all the include paths for external code just to compile the documentation.

**Comment by Livingwind** [comment]
That being said, I don't think the expected behavior is just putting `int` in place of the unknown type.

**Comment by jnikula** [comment]
Not that it's any consolation to you, but we don't invent the `int` here. We just use the AST clang produces, and it's obviously incomplete when the type is incomplete. [GIGO](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out) if you will.

We'll need to see if there's anything simple we can do in this case. But to set the expectations right, 1) for the time being we still have more important things to cover in the "happy day scenarios", and 2) any potential fix will need to be fairly simple - I don't want to overcomplicate things, at least not until the internals of the parser parts have matured.

Regardless, thanks for the feedback from the real world, we need that.

**Comment by jnikula** [comment]
> Is there no way to just print the symbol even though the compiler fails? It seems like a big pain to add all the include paths for external code just to compile the documentation.
> 
> It's also a bit strange that 1 linkage failure within a structure makes the structure anonymous rather than providing it's typedef'ed name.

I've split this part into issue #13.

**Comment by mv0** [comment]
Apologies, hand't had the time to test this more. Probably this should be another ticket (briefly tested branch you mention @BrunoMSantos). 

If we are on the same "struct" subject can someone confirm this behaviour:

Assuming you have the following:
```c
/**
 * this is documentation for struct
 */
struct ops {
        /**
         * write the data
         *
         * @param data a pointer to a buf
         * @param data_len the amount to write, in bytes
         */
        void (*write)(struct ops *ops, const char *data, size_t data_len);
        /**
         * read the data
         *
         * @param data a pointer to a buf
         * @param data_len the amount to write, in bytes
         */
        void (*read)(struct ops *ops, const char *data, size_t data_len);
};
```

Which generates (with both master and your branch): 

```c

.. c:type:: struct ops

   this is documentation for struct


   .. c:member:: void (*)(struct ops *, const char *, int) write

      write the data

      @param data a pointer to a buf
      @param data_len the amount to write, in bytes


   .. c:member:: void (*)(struct ops *, const char *, int) read

      read the data

      @param data a pointer to a buf
      @param data_len the amount to write, in bytes
```

The name of the pointer to function is suffixed, but let say that wouldn't be such an issue but the args for it are not included in the declaration. Imagine that `data` hand't had `const char *` type but something more opaque like a typedef and the same for int. Imagine you have 4, 5 args. Imagine the code is new to you. How do you make out what types the args are? 

@jnikula is this a feature or a bug? @BrunoMSantos?

**Comment by BrunoMSantos** [comment]
@mv0, I'll have to check the compiled output to be sure, but there's probably something wrong here that deserves its own ticket. This is totally unrelated to anonymous types though, so it's not a surprise that the `anonym` branch doesn't help. I also don't think we've tested function pointers so far, so it's not surprising if it's broken.

As for this ticket, it looked like it was about anonymous types, but then the discussion turned to handling of undefined types, which is already a separate issue.  I agree with @jnikula, there's not much we can do about that, but we can put it in a separate issue. Personally I don't see the value. 

That leaves only the anonymous stuff and, for what it's worth, I'd still like a clear indication of what you guys would expect relative to anonymous types. What I originally had in mind will not work without changing upstream as far as I know (feature n. 3 from my previous reply), but I'd like to have a clear indication (pseudocode examples if possible) of what you'd expect before I do anything else.
